datastore/threads/threadproxyds.nim

```diff
@@ -51,9 +53,43 @@ type
     queryLock: AsyncLock      # global query lock, this is only really \
                               # needed for the fsds, but it is expensive!
 
+proc addrOf*[T](ctx: ref TaskCtx[T]): ptr TaskCtx[T] =
+  result = cast[ptr TaskCtx[T]](ctx)
+  echo "ADDR_OF: ", result.pointer.repr
+
+proc new*[T](
+    ctx: typedesc[TaskCtx[T]],
+    ds: Datastore,
+): ref TaskCtx[T] =
+  result = (ref TaskCtx[T])()
+  result.ds = unsafeAddr(ds) ##\
+    ## doing this appears to break things. previously it was using `addr(ds)`
```

Comment:
You really don't want to do this, because there is no way for the compiler to trace the lifetime of `ds` ref type, that is why you cant use `addr` in this context and you are forced to break out of the compiler's safety with `unsafeAddr`. 

On top of that, you're turning `TaskCtx` into a `ref` for a reason, because we actually want to control the lifetime of the `ctx` and constrain it to the stack's lifetime.

Reply:
> You really don't want to do this, because there is no way for the compiler to trace the lifetime of ds ref type, that is why you cant use addr in this context and you are forced to break out of the compiler's safety with unsafeAddr.

Yes, it's not a good way to approach it, but it highlights the challenge here. I think just changing it the type of `ds: ptr Datastore` to `ds: Datastore` would be better. However, that also breaks the GCs as I've been finding in my testing.

That lifetime analysis isn't quite right as in general it's not safe to 
use `addr` here either. It's just *slightly* more safe since there's a guaranteed stack address. The compiler can't determine the lifetime or use it to track the datastore's life when you use `addr` anymore than with `unsafeAddr`. They both become `ptr Datastore` which the GC doesn't understand. It sorta works because the `var ds` here get's captured into the Future whose lifetime generally exceeds the task, but not always.

At least the Future's lifetime exceeds it in the normal control flow, but passing it to the thread-tasks still bypasses that. The situation is similar to if you try to capture a `var ds` into a closure - the compiler will error because doing so isn't safe even in a non-ref object. So just using a `ds: Datastore` and then an additional `GC_ref` would properly increase the ref count of the `ds` which could then be decreased when both sides are done.